test/types/component-test.ts

```diff
@@ -1,39 +1,96 @@
+import { expectTypeOf } from 'expect-type';
 import * as gc from '@glimmer/component';
-import { hasExactKeys } from './utils';
 
 const Component = gc.default;
 
-hasExactKeys<{
-  default: unknown;
-}>()(gc);
-
-// $ExpectType typeof GlimmerComponent
-gc.default;
+expectTypeOf(gc).toHaveProperty('default');
+expectTypeOf(gc.default).toEqualTypeOf<typeof Component>();
 
 type Args = {
   foo: number;
 };
 
-const component = new Component<Args>({}, { foo: 123 });
+const componentWithLegacyArgs = new Component<Args>({}, { foo: 123 });
+expectTypeOf(componentWithLegacyArgs).toHaveProperty('args');
+expectTypeOf(componentWithLegacyArgs).toHaveProperty('isDestroying');
+expectTypeOf(componentWithLegacyArgs).toHaveProperty('isDestroyed');
+expectTypeOf(componentWithLegacyArgs).toHaveProperty('willDestroy');
+expectTypeOf(componentWithLegacyArgs.args).toEqualTypeOf<Readonly<Args>>();
+expectTypeOf(componentWithLegacyArgs.isDestroying).toEqualTypeOf<boolean>();
+expectTypeOf(componentWithLegacyArgs.isDestroyed).toEqualTypeOf<boolean>();
+expectTypeOf(componentWithLegacyArgs.willDestroy).toEqualTypeOf<() => void>();
+
+interface ArgsOnly {
+  Args: Args;
+}
+
+const componentWithArgsOnly = new Component<ArgsOnly>({}, { foo: 123 });
+expectTypeOf(componentWithArgsOnly.args).toEqualTypeOf<Readonly<Args>>();
+
+interface ElementOnly {
+  Element: HTMLParagraphElement;
+}
+
+const componentWithElOnly = new Component<ElementOnly>({}, {});
+
+// We cannot check on toEqualTypeOf here b/c EmptyObject is intentionally not
+// public.
```

Comment:
Could this cause issues with declaration emit? I'm not sure how likely this actually is to come up in practice, but as a super contrived example, if I were to write something like:

```ts
class MyComponent extends Component<{}> {
  foo = this.args;
}
```

I think `tsc` would fail trying to produce a type annotation for `foo` for a `.d.ts` file, wouldn't it?

Reply:
Good call. I’ll export this with an `@internal` annotation… and we’ll need to treat its *location* as public API for imports so existing declarations don’t break.

Reply:
Resolved via 6fc4a17!